2.6 jsonView should throw exception if json_encode fails

robmcvey opened the pull request:
Closes https://github.com/cakephp/cakephp/issues/3141

Seems this issue lay dormant, but solution given by @dereuromark in https://github.com/dereuromark/cakephp/commit/7b6238613de79bb0135d557a182c663ad3e9f8a1 seems a good fix.

I've added extra catch for if `json_last_error()` nor `json_last_error_msg()` are present.

robmcvey committed:
Assert CakeException throw when json_encode fails

robmcvey committed:
Handle json_encode failure

robmcvey committed:
Coding standards

robmcvey committed:
Encoding false is OK - should return string null

dereuromark commented on lib/Cake/View/JsonView.php:
```diff
@@ -145,10 +149,22 @@ protected function _serialize($serialize) {
+				$error = __('JSON encoding failed: Error code %s', json_last_error());
+			}
+			throw new CakeException($error);
+		} else if ($json === false) {
```
should be elseif

robmcvey committed:
coding standards elseif

dereuromark commented:
I don't think we should translate those internal server errors. What do others think?

robmcvey committed:
suppress warnings so test runner doesn't fart

markstory commented:
There are still a [few PHPCS errors](https://travis-ci.org/cakephp/cakephp/jobs/54272911).

@dereuromark Agreed, those error messages do not need to be translated.

ADmad commented on lib/Cake/View/JsonView.php:
```diff
@@ -145,10 +149,22 @@ protected function _serialize($serialize) {
 		}
 
-		return json_encode($data);
+		if (function_exists('json_last_error') && json_last_error() !== JSON_ERROR_NONE) {
```
You can backport the [json_last_error()](https://github.com/cakephp/cakephp/blob/3.0/src/basics.php#L132) added to basics.php in 3.0.

robmcvey committed:
Don't translate errors

robmcvey committed:
add throws tag to function comment

robmcvey committed:
fallback implementation of json_last_error_msg

robmcvey committed:
Use fall back json_last_error_msg

robmcvey commented:
Suppressing the json_encode warnings with `@` fails code sniffer, but without PHPUnit throws `PHPUnit_Framework_Error_Warning` before we can catch anything in the test case.

What would be the best approach to get around this? set_error_handler() ?

lorenzo commented:
Use `set_error_handler` in the test

robmcvey committed:
add a custom error handler for tests of json_encode

robmcvey committed:
Remove tabs

robmcvey committed:
Assert exception is thrown, reset error handler in tearDown

robmcvey committed:
Tickle Travis

markstory commented on lib/Cake/Test/Case/View/JsonViewTest.php:
```diff
@@ -28,12 +28,27 @@
+ **/
+	public function tearDown() {
+		parent::tearDown();
+		restore_error_handler();
```
You should only restore the error handler when you change it. Otherwise you can end up popping off phpunits error handler.

robmcvey commented:
Ah OK thanks, would this be the preferred way?

```php
try {
	$View->render();
	$this->fail('Failed asserting that exception of type "CakeException" is thrown.');
} catch (CakeException $e) {
	restore_error_handler();
	return;
}

markstory commented:
@robmcvey That looks good to me. I would normally do some assertions in the catch block too.

robmcvey committed:
assert correct exception msg thrown

robmcvey committed:
Assert exception msg contains UTF-8

The different versions of PHP throw several various messages for UTF-8
sequences, so this just performs a basic regex check

robmcvey committed:
Remove whitespace

lorenzo commented:
Good job @robmcvey :)
